src/librustc_data_structures/obligation_forest/mod.rs

```diff
@@ -42,34 +66,40 @@ pub struct ObligationForest<O, T> {
     /// at a higher index than its parent. This is needed by the
     /// backtrace iterator (which uses `split_at`).
     nodes: Vec<Node<O>>,
-    trees: Vec<Tree<T>>,
-    snapshots: Vec<usize>,
+    done_cache: FnvHashSet<O::Predicate>,
```

Comment:
these fields could use some comments indicating the criteria by which a predicate is added to these various caches etc